### PR TITLE
[@types/express-serve-static-core] 🏷️ Add `Promise<void>` to `RequestHandler`'s return type

### LINKED ISSUE #50871: [@types/express-serve-static-core] async handlers typings
I tried bumping `@types/express-serve-static-core` to the latest version but the [changes to the return type of handlers](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/7a41e02d90d3c0cc02104272a28ac4d89b5d64d4#diff-5962769e8693df44c9e1083881fb1da149a67262026db939b7430536d514e7af) (to `void`) introduced a typing error for async handlers.

Should we instead go with a `void | Promise<void>` return type?

### PR DESCRIPTION
**Pull request by NatoBoram**
* Fixes #50871
* Adds `@typescript-eslint/no-misused-promises` to `@types/express-serve-static-core`'s ESLint config
* Fixes tests that returns something from an Express handler, something that's not supposed to be possible in the first place as handlers are supposed to return `void` (and now `void | Promise<void>`)
* Closes https://github.com/coderabbitai/DefinitelyTyped/pull/1

### CODE REVIEW

File: types/mock-req-res/mock-req-res-tests.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import { Request, RequestHandler, Response } from "express";
 import { mockRequest, mockResponse } from "mock-req-res";
 
 const handler: RequestHandler = (req: Request, res: Response) => {
-    return res.status(200).json(`Hello from handler with an originalUrl value of '${req.originalUrl}'`);
+    return void res.status(200).json(`Hello from handler with an originalUrl value of '${req.originalUrl}'`);
```
Comment (NatoBoram) [inline types/mock-req-res/mock-req-res-tests.ts:5]: This `return` and many others are entirely optional, but are kept as `return void` for demonstration purposes.

File: types/express-serve-static-core/.eslintrc.json [hunk 2]
```diff
@@ -2,6 +2,7 @@
     "rules": {
         "@definitelytyped/no-type-only-packages": "off",
         "@typescript-eslint/naming-convention": "off",
-        "@typescript-eslint/no-empty-interface": "off"
+        "@typescript-eslint/no-empty-interface": "off",
+        "@typescript-eslint/no-misused-promises": "error"
```
Comment (NatoBoram) [inline types/express-serve-static-core/.eslintrc.json:6]: This is the linting issue that's giving troubles in #50871 and the reason for this PR. Using typescript-eslint's `strict-type-checked` is a good way to ensure code quality/safety/maintainability and this rule is part of it.

File: types/express-serve-static-core/express-serve-static-core-tests.ts [hunk 3]
```diff
@@ -323,3 +323,8 @@ app.get("/:readonly", req => {
     // @ts-expect-error
     req.xhr = true;
 });
+
+// Starting with Express 5, route handlers and middleware that return a
+// `Promise` will call `next(value)` automatically when they reject or throw an
+// error.
```
Comment (NatoBoram) [inline types/express-serve-static-core/express-serve-static-core-tests.ts:329]: Copy/pasted from Express' documentation.

**Comment by NatoBoram** [comment]
Would be nice if I could test locally without filling my swap and crashing my system tbh

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/mongoose-aggregate-paginate-v2/mongoose-aggregate-paginate-v2-tests.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against cb3428aac23d5031469f8e01906834f0f06d8792
</p>

**Review by cw-alexcroteau — APPROVED** [review]

**Review by acrilex1 — APPROVED** [review]

File: types/express-serve-static-core/.eslintrc.json (hunk 2 above)
Comment (sheetalkamat) [inline types/express-serve-static-core/.eslintrc.json:6]: Pls do not add any more rules to eslint. The good thing to do here would be to remove all the overrides here and instead fix the errors if any,

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/express-serve-static-core/.eslintrc.json (hunk 2 above)
Comment (NatoBoram) [reply types/express-serve-static-core/.eslintrc.json]: I have reverted my changes to this file in f9f4929abf so that this pull request no longer tests its own change, as per your request.

Fixing the overrides would be out of scope for this pull request.

**Review by cw-alexcroteau — APPROVED** [review]

**Review by acrilex1 — APPROVED** [review]

File: types/express-serve-static-core/index.d.ts [hunk 4]
```diff
@@ -61,7 +61,7 @@ export interface RequestHandler<
         req: Request<P, ResBody, ReqBody, ReqQuery, LocalsObj>,
         res: Response<ResBody, LocalsObj>,
         next: NextFunction,
-    ): void;
+    ): void | Promise<void>;
```
Comment (gabritto) [inline types/express-serve-static-core/index.d.ts:64]: Could you clarify why this change was needed? I read https://github.com/DefinitelyTyped/DefinitelyTyped/issues/50871 but I still don't understand why this change is needed.

**Review by gabritto — CHANGES_REQUESTED** [review]
Requested clarification.

File: types/express-serve-static-core/index.d.ts (hunk 4 above)
Comment (NatoBoram) [reply types/express-serve-static-core/index.d.ts:64]: If you're using TypeScript-ESLint's [recommended-type-checked](https://typescript-eslint.io/users/configs/#recommended-type-checked), which comes with [no-misused-promises](https://typescript-eslint.io/rules/no-misused-promises/), you get an error whenever you create an Express endpoint with an async function.

```ts
import express from "express"

const app = express()

app.post("/", async () => {
	await new Promise<void>(resolve => setTimeout(resolve, 1000))
})
```

```sh
/home/nato/Code/github.com/coderabbitai/rabbits-lab/src/express/promise.ts
  5:15  error  Promise returned in function argument where a void return was expected  @typescript-eslint/no-misused-promises

✖ 1 problem (1 error, 0 warnings)
```

This is because the type definition for `RequestHandler` incorrectly asserts that no async function can ever be given to `app.post()` and that whatever is using `RequestHandler` should always assume that the return type of that handler is exactly `void` and nothing else.

That is not true, since Express itself is able to handle promise rejections from `RequestHandler`.

The type needs to be changed so that linters no longer throw an error in figuratively every usage of Express.

**Comment by NatoBoram** [comment]
@typescript-bot no new commits needed to be pushed

File: types/express-serve-static-core/index.d.ts (hunk 4 above)
Comment (gabritto) [reply types/express-serve-static-core/index.d.ts:64]: Ok, I now understand the part about lint errors, thank you.

> That is not true, since Express itself is able to handle promise rejections from RequestHandler.

Is that documented somewhere?

Comment (NatoBoram) [reply types/express-serve-static-core/index.d.ts:64]: Yes! The code comment I placed was copied from [expressjs.com/en/guide/error-handling](https://expressjs.com/en/guide/error-handling.html#:~:text=starting%20with%20express%205%2C%20route%20handlers%20and%20middleware%20that%20return%20a%20promise%20will%20call%20next(value)%20automatically%20when%20they%20reject%20or%20throw%20an%20error.)

There's also another example usage of an async function in a `RequestHandler` in [expressjs.com/en/guide/writing-middleware](https://expressjs.com/en/guide/writing-middleware.html#:~:text=note%20how%20next()%20is%20called%20after%20await)

**Review by gabritto — APPROVED** [review]

**Comment by NatoBoram** [comment]
Ready to merge

**Comment by AviVahl** [comment]
While express 5 would accept and handle promises, this PR changes express 4's types.
Also, express@5 final release isn't out, is it?

Are we trying to force everyone using typescript-eslint's type-checking rules with express 4 to write the prefix `void` everywhere?

I don't think this change represent the current state of express, and I'm quite sure typescript-eslint wouldn't warn me when I wrongfully pass in an async fn as an handler that can possibly reject, with nothing handling that rejection.

**Comment by gabritto** [comment]
Sorry for my oversight. @NatoBoram I reverted this PR, you could recreate it but change the types of `express-serve-static-core` to version 5 instead.

**Comment by NatoBoram** [comment]
Oh, I didn't realize I was the one that had to change version number, I thought this would be handled by the team.

I've re-created it with `v5.0.9999`

* https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69846

**Comment by karlhorky** [comment]
> Also, express@5 final release isn't out, is it?

`express@5.0.0` is now published to the npm `next` tag:

- https://github.com/expressjs/express/pull/2237#issuecomment-2340006831
